File: ios/EarthCore/Sources/EarthCore/Chain/Fees.swift
import BigInt
import Foundation

/// The fee a transaction must offer, derived from its gas limit.
///
/// A Cosmos node rejects a transaction whose fee is below
/// `ceil(gas_limit * minimum-gas-prices)`, which makes the fee a function of
/// the gas limit rather than a flat amount. Both apps had it as a flat "2000" —
/// the fee for the 400,000-gas default — while `Personhood.registerGasLimit`
/// had been raised to 3,000,000 for headroom. Every registration was rejected
/// before it ran:
///
///     insufficient fees; got: 2000uerth required: 15000uerth
///
/// Gas headroom is not free. It costs fee, proportionally, whether or not the
/// gas is used: the fee is what you offer, and an unused remainder is not
/// refunded. Raising a gas limit without raising the fee breaks the
/// transaction; raising both makes it more expensive.
///
/// The price comes from the node rather than a constant, because
/// `minimum-gas-prices` is per-node configuration and not a chain parameter —
/// an operator changes it on a restart, with no governance vote and no signal
/// to clients.
public enum Fees {

    /// Used until ``prime(rest:)`` succeeds, and if it never does.
    ///
    /// Matches what `deploy/akash/deploy.yaml` sets on the validator
    /// (`MIN_GAS_PRICES=0.005uerth`). Being wrong here is recoverable — the
    /// node names the required amount in its rejection — whereas refusing to
    /// build a transaction at all is not.
    static let fallbackPrice = Decimal(string: "0.005")!

    private static let lock = NSLock()
    private static var cached: Decimal?

    /// The fee, in uerth, for `gas`.
    ///
    /// Synchronous and non-throwing on purpose: it is read while building a
    /// confirmation sheet and while working out a spendable balance, neither of
    /// which can await a network round trip. ``prime(rest:)`` fills the cache;
    /// until then this answers with ``fallbackPrice``.
    public static func forGas(_ gas: UInt64) -> String {
        var raw = (readCache() ?? fallbackPrice) * Decimal(gas)
        var rounded = Decimal()
        // Rounded up: the node compares against a ceiling, so truncating
        // produces a fee one uerth short and a rejection that reads as a
        // rounding mystery.
        NSDecimalRound(&rounded, &raw, 0, .up)
        return NSDecimalNumber(decimal: rounded).stringValue
    }

    /// As ``forGas(_:)``, for callers that want to compare against a balance.
    public static func forGasValue(_ gas: UInt64) -> BigInt {
        BigInt(forGas(gas)) ?? BigInt(0)
    }

    /// Reads `minimum-gas-prices` from the node and caches it for the process.
    ///
    /// Safe to call repeatedly and cheap once primed: the value cannot change
    /// under a running node. Never throws — a node that cannot be reached at
    /// launch must not stop the app from building transactions, and the
    /// fallback is the right answer on this chain anyway.
    public static func prime(rest: EarthRest = EarthRest()) async {
        if readCache() != nil { return }
        guard let price = try? await fetch(rest: rest) else { return }
        writeCache(price)
    }

    // Locking lives in these two, deliberately: NSLock is unavailable from an
    // async context (a hard error in the Swift 6 language mode), and holding a
    // lock across an await is how you deadlock anyway. Neither of these awaits.
    private static func readCache() -> Decimal? {
        lock.lock()
        defer { lock.unlock() }
        return cached
    }

    private static func writeCache(_ price: Decimal) {
        lock.lock()
        defer { lock.unlock() }
        cached = price
    }

    /// The endpoint answers `"0.005000000000000000uerth"` — a decimal with the
    /// denom appended, so the denom comes off before parsing. Only uerth is
    /// accepted: this app pays fees in nothing else, and a node quoting another
    /// denom is a misconfiguration the fallback handles better than a silently
    /// wrong number would.
    private static func fetch(rest: EarthRest) async throws -> Decimal? {
        let json = try await rest.get("/cosmos/base/node/v1beta1/config")
        guard let raw = json["minimum_gas_price"].string, raw.hasSuffix("uerth") else {
            return nil
        }
        let digits = raw.prefix { $0.isNumber || $0 == "." }
        guard !digits.isEmpty, let price = Decimal(string: String(digits)) else { return nil }
        return price
    }
}

File: android/app/src/main/java/com/example/earthwallet/chain/Fees.kt
package network.erth.wallet.chain

import org.json.JSONObject
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * The fee a transaction must offer, derived from its gas limit.
 *
 * A Cosmos node rejects a transaction whose fee is below
 * `ceil(gas_limit * minimum-gas-prices)`. That makes the fee a function of the
 * gas limit, not a flat amount — which is exactly the assumption this app used
 * to get wrong. Personhood.REGISTER_GAS_LIMIT was raised to 3,000,000 for
 * headroom while the fee stayed at a flat 2,000 uerth, and every registration
 * was rejected before it ran:
 *
 *     insufficient fees; got: 2000uerth required: 15000uerth
 *
 * Gas headroom is not free. It costs fee, proportionally, whether or not the
 * gas is used — the fee is what you offer, and an unused remainder is not
 * refunded. Raising a gas limit without raising the fee breaks the transaction;
 * raising both makes it more expensive.
 *
 * The price comes from the node rather than a constant. `minimum-gas-prices` is
 * per-node configuration, not a chain parameter, so an operator can change it
 * on a restart with no governance vote and no warning to clients. The value is
 * cached for the process: it cannot change under a running node.
 */
object Fees {

    /**
     * Fallback for when the node cannot be reached or answers with something
     * unparseable. Matches the value deploy/akash/deploy.yaml sets on the
     * validator (MIN_GAS_PRICES=0.005uerth). Being wrong here is recoverable —
     * the transaction is rejected with the required amount in the error — while
     * refusing to build a transaction at all is not.
     */
    private val FALLBACK_PRICE: BigDecimal = BigDecimal("0.005")

    @Volatile
    private var cached: BigDecimal? = null

    /**
     * The fee, in uerth, for [gasLimit] gas.
     *
     * Rounded up: the node compares against a ceiling, so truncating produces a
     * fee one uerth short and a rejection that looks like a rounding mystery.
     */
    fun forGas(gasLimit: Long): Long =
        (cached ?: FALLBACK_PRICE)
            .multiply(BigDecimal(gasLimit))
            .setScale(0, RoundingMode.CEILING)
            .toLong()

    /** As [forGas], for the call sites that want the amount as a string. */
    fun forGasString(gasLimit: Long): String = forGas(gasLimit).toString()

    /**
     * Fetches the price and caches it. Call from a background thread.
     *
     * Deliberately separate from [forGas], which must never touch the network:
     * fees are read from composables to show a confirm sheet and to work out a
     * spendable balance, and a network call on the main thread is an instant
     * NetworkOnMainThreadException. So [forGas] answers from cache or falls
     * back, and this is what fills the cache — from app startup and from the IO
     * thread a broadcast already runs on.
     *
     * Cheap to call repeatedly: it returns immediately once primed, and
     * minimum-gas-prices cannot change under a running node.
     */
    fun prime() {
        if (cached != null) return
        cached = fetch() ?: return   // leave unset on failure so a later call retries
    }

    /**
     * Reads minimum-gas-prices from the node.
     *
     * The endpoint answers "0.005000000000000000uerth" — a decimal with the
     * denom appended, so the denom has to come off before parsing. Only uerth
     * is handled: this app pays fees in nothing else, and a node quoting a
     * different denom is a misconfiguration the fallback handles better than a
     * silently wrong number would.
     */
    private fun fetch(): BigDecimal? = runCatching {
        val (code, body) = EarthRest.get("/cosmos/base/node/v1beta1/config")
        if (code != 200) return null
        val raw = JSONObject(body).optString("minimum_gas_price").ifEmpty { return null }
        val digits = raw.takeWhile { it.isDigit() || it == '.' }
        if (digits.isEmpty() || !raw.endsWith("uerth")) return null
        BigDecimal(digits)
    }.getOrNull()
}
